compiler/rustc_lint/src/internal.rs

```diff
@@ -364,54 +367,133 @@ declare_lint_pass!(Diagnostics => [UNTRANSLATABLE_DIAGNOSTIC, DIAGNOSTIC_OUTSIDE
+        // Is the parent method marked with `#[rustc_lint_diagnostics]`?
+        let mut parent_has_attr = false;
+        for (hir_id, _parent) in cx.tcx.hir().parent_iter(expr.hir_id) {
+            if let Some(owner_did) = hir_id.as_owner()
+                && cx.tcx.has_attr(owner_did, sym::rustc_lint_diagnostics)
             {
-                found_diagnostic_message = true;
+                parent_has_attr = true;
                 break;
             }
         }
-        debug!(?found_diagnostic_message);
-        if !found_parent_with_attr && !found_diagnostic_message {
-            cx.emit_span_lint(UNTRANSLATABLE_DIAGNOSTIC, span, UntranslatableDiag);
+
+        // Calls to `#[rustc_lint_diagnostics]`-marked functions should only occur:
+        // - inside an impl of `IntoDiagnostic`, `AddToDiagnostic`, or `DecorateLint`, or
+        // - inside a parent function that is itself marked with `#[rustc_lint_diagnostics]`.
+        //
+        // Otherwise, emit a `DIAGNOSTIC_OUTSIDE_OF_IMPL` lint.
+        if has_attr && !parent_has_attr {
+            let mut is_inside_appropriate_impl = false;
+            for (_hir_id, parent) in cx.tcx.hir().parent_iter(expr.hir_id) {
+                debug!(?parent);
+                if let Node::Item(Item { kind: ItemKind::Impl(impl_), .. }) = parent
+                    && let Impl { of_trait: Some(of_trait), .. } = impl_
+                    && let Some(def_id) = of_trait.trait_def_id()
+                    && let Some(name) = cx.tcx.get_diagnostic_name(def_id)
+                    && matches!(
+                        name,
+                        sym::IntoDiagnostic | sym::AddToDiagnostic | sym::DecorateLint
+                    )
+                {
+                    is_inside_appropriate_impl = true;
+                    break;
+                }
+            }
+            debug!(?is_inside_appropriate_impl);
+            if !is_inside_appropriate_impl {
+                cx.emit_span_lint(DIAGNOSTIC_OUTSIDE_OF_IMPL, span, DiagOutOfImpl);
+            }
+        }
+
+        // Calls to methods with an `impl Into<{D,Subd}iagnosticMessage>` parameter must be passed
+        // an arg with type `{D,Subd}iagnosticMessage` or `impl Into<{D,Subd}iagnosticMessage`.
+        // Otherwise, emit an `UNTRANSLATABLE_DIAGNOSTIC` lint.
+        //
+        // This is imprecise. It doesn't match args to parameters exactly, and could mis-analyze a
+        // function like `fn foo(msg1: DiagnosticMessage, msg2: impl Into<DiagnosticMessage>)`, but
+        // there are no such functions in practice, so this is good enough for now.
+        if num_impl_into_diagnostic_message_params > 1 {
+            cx.tcx.dcx().span_bug(span, "multiple `impl Into<{D,Sub}diagnosticMessage>` params")
+        } else if num_impl_into_diagnostic_message_params == 1 {
+            let mut has_diagnostic_message_arg = false;
+            let mut has_impl_into_diagnostic_message_arg = false;
+            for ty in call_tys {
```

Comment:
How about saving the indices of Into<DiagnosticMessage> params above so you can check just these params? It would also handle multiple parameters properly then.

Reply:
I was advised to use `predicates_of`, but I'm still puzzled by it. For a call to a `Diag` method like this:
```rust
    pub fn struct_span_err(
        &self,
        span: impl Into<MultiSpan>,
        msg: impl Into<DiagnosticMessage>,
    ) -> Diag<'_>;
```
`predicates_of` returns this:
```
preds: [
    Binder { value: TraitPredicate(<impl Into<MultiSpan> as std::marker::Sized>, polarity:Positive), bound_vars: [] },
    Binder { value: TraitPredicate(<impl Into<DiagnosticMessage> as std::marker::Sized>, polarity:Positive), bound_vars: [] },
    Binder { value: TraitPredicate(<impl Into<MultiSpan> as std::convert::Into<rustc_errors::MultiSpan>>, polarity:Positive), bound_vars: [] },
    Binder { value: TraitPredicate(<impl Into<DiagnosticMessage> as std::convert::Into<rustc_errors::DiagnosticMessage>>, polarity:Positive), bound_vars: [] }
]
```
So the earlier code looks through these and sees the the last element and knows that one of the args is `impl Into<DiagnosticMessage>`. But it doesn't know which one because the indices here don't match the parameter indices.

What I really expected when I started was to instead look up the type signature of `struct_span_err` and go from there. Then I could compare each parameter from the signature against the corresponding argument from the call site. How would I do that? It seems like it might be easier.

Reply:
Just to reiterate, I'm new to this kind of linting/typechecking code, so things that would be obvious to a more experienced person might be non-obvious to me! I'm just groping my way along.

Reply:
There are two different kinds of "parameter" indices, the type parameter indices (the identity of `ty::Param`) and the function parameter indices (from the `fn_sig()`).
What we want to know for the check is which *function* parameters are  `Into<DiagnosticMessage>`, so we can check those function parameters.
To check whether a function parameter type is `Into<DiagnoticMessage>`, we have two parts. First, the type must be a `ty::Param` (as it's a generic parameter) *and* there must be a trait predicate where *this type* is the self type and the trait is `DiagnosticMessage`.

So I'd propose this pseudocode:
```rust
fn params_that_are_into_diagnostic_message(def_id) -> Vec<bool> {
  let sig = fn_sig(def_id);
  let predicates = predicates_of(def_id);
  let mut is_into_diagnostics_message = vec![false; sig.params().len()]
  for i in 0..sig.params().len() {
    let ty = sig.params()[i];
    if let ty::Param(..) = ty {
      if predicates.any(|p| p.self_ty == ty && p.trait == sym::DiagnosticsMessage) {
        is_into_diagnostics_message[i]  = true;
      }
    }
  }
} 
```

Reply:
Aha, the `fn_sig` query was the key thing I didn't know about. Once I had that everything fell into place. Thank you!